initial checkin for core / tests

m-nash opened the pull request:
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [x] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/master/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running generate.ps1. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version,

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

m-nash committed:
initial checkin for core / tests

m-nash committed:
Add readme.md file

m-nash committed:
Fix header in readme

m-nash committed:
remove autorest until we convert off hand written types

m-nash committed:
update azure.core reference
remove sets to tags
ignore tests which require credentials for now

m-nash committed:
Add the proto client which will be used until the autogen work is complete

m-nash committed:
Add missing readmes for proto client

m-nash commented:
@pakrym here is the apiView you requested https://apiview.dev/Assemblies/Review/393c643a31a344e2805f738350665802 just know this is still a WIP which is why its going into a feature branch vs master so some comments on the apiView will be taken as follow up

weshaggard commented on eng/Packages.Data.props:
```diff
@@ -64,14 +64,16 @@
     <PackageReference Update="Microsoft.CSharp" Version="4.5.0" />
-
+    <PackageReference Update="System.Collections.Immutable" Version="1.3.1" />
+    <PackageReference Update="System.Linq.Async" Version="5.0.0" />
```
@pakrym I know there has been some discussion about AsyncEnumerables but I don't recall what the outcome was. Is there any reason we shouldn't use this library?

pakrym replied:
We were able to implement all AsyncEnumerable features in other libraries without using the `System.Linq.Async` wonder why it's required in this case.

m-nash replied:
this line here https://github.com/m-nash/azure-proto-sdk/blob/master/Azure.ResourceManager.Core/ResourceOperationsBase.cs#L192 is not possible without System.Linq.Async.  There are other ways to implement it which I am open to if we don't want to add this dependency.

weshaggard replied:
If that is the only code that uses it then I definitely think we should avoid this dependency and implement that directly as it isn't worth taking that dependency.

pakrym replied:
You can turn that line into a foreach loop:


```C#
await foreach (var p in pageableProvider)
{
	if (string.Equals(p.Namespace, resourceType?.Namespace, StringComparison.InvariantCultureIgnoreCase))
	{
	  resourcePageableProvider = p;
	  break
	}
}

```

m-nash replied:
Added an ADO item for this

weshaggard commented on sdk/resourcemanager/Azure.ResourceManager.Core/src/Azure.ResourceManager.Core.csproj:
```diff
@@ -0,0 +1,22 @@
+      This is a beta preview vesion.  This version uses a next-generation code generator that introduces important breaking changes, but also new features (such as intuitive authentication, custom HTTP pipeline, distributed tracing and much more).
+    </Description>
+    <PackageTags>azure;management;resource</PackageTags>
+    <NoWarn>$(NoWarn);AZC0008;AZC0001;AZC0107;CA2214;CA1036;CA1067;CA1065;SA1028</NoWarn>
```
Lets make sure we have some comments here or issues tracking the fixing of these warning.

m-nash replied:
I have an ADO item to follow up on removing all of those

weshaggard commented on sdk/resourcemanager/Proto.Client/authorization/lib/Azure.ResourceManager.Authorization.xml:
```diff
@@ -0,0 +1,1804 @@
+<?xml version="1.0"?>
```
We should not be checking in binaries into our repo. We should either commit the source and build this or pull it from a package manager.

weshaggard commented on sdk/resourcemanager/Proto.Client/authorization/README.MD:
```diff
@@ -0,0 +1,3 @@
+# Proto Authorization
```
What is the ultimate goal for this Proto.Client? Is it is going to become a library we ship? Is it just going to be for testing or is it going to be a sample? If it is going to be a test or sample we should put it under the tests or samples folder. I

m-nash replied:
This is temporary until the autogen work is done it allows us to test our arm core in a working end to end state

weshaggard commented:
/azp prepare-pipelines

weshaggard commented:
/azp run prepare-pipelines

pakrym commented on eng/Packages.Data.props:
```diff
@@ -64,14 +64,16 @@
+    <PackageReference Update="Microsoft.Bcl.AsyncInterfaces" Version="1.1.1" />
     <PackageReference Update="Microsoft.CSharp" Version="4.5.0" />
-
+    <PackageReference Update="System.Collections.Immutable" Version="1.3.1" />
```
We shouldn't expose any immutable collections in the API and they shouldn't be required for implementaitons.

m-nash replied:
Added an ADO item to track this

pakrym commented on eng/Packages.Data.props:
```diff
@@ -64,14 +64,16 @@
     <PackageReference Update="System.Text.Json" Version="4.6.0" />
     <PackageReference Update="System.ValueTuple" Version="4.5.0" />
-    <PackageReference Update="Microsoft.Bcl.AsyncInterfaces" Version="1.0.0" />
+    <PackageReference Update="Microsoft.Bcl.AsyncInterfaces" Version="1.1.1" />
```
Any particular reason for the version update?

m-nash replied:
Needed for System.Linq.Async

pakrym replied:
You can let whoever uses `System.Linq.Async` to pull the new version transitively and not update the global one.

pakrym commented on eng/Packages.Data.props:
```diff
@@ -64,14 +64,16 @@
     <PackageReference Update="System.Diagnostics.DiagnosticSource" Version="4.6.0" />
     <PackageReference Update="System.Reflection.TypeExtensions" Version="4.5.1" />
-    <PackageReference Update="System.Threading.Tasks.Extensions" Version="4.5.2" />
+    <PackageReference Update="System.Threading.Tasks.Extensions" Version="4.5.4" />
```
Any particular reason for the version update?

m-nash replied:
I think this was required by System.Linq.Async

m-nash commented:
@pakrym adding the example RP APIView that uses the new Azure.ResourceManager.Core.  Again this is WIP so not all comments will be address immediately but we will track them in ADO
https://apiview.dev/Assemblies/Review/12422a8f348249a18e2ff05e6ae776e5

weshaggard commented:
/azp run net - resourcemanager - ci

m-nash committed:
update to require linq 4.0.0 so that don't have to upgrade bcl

m-nash committed:
add changlog for CI

weshaggard commented on eng/Packages.Data.props:
```diff
@@ -71,6 +71,8 @@
     <PackageReference Update="Microsoft.Bcl.AsyncInterfaces" Version="1.0.0" />
     <PackageReference Update="Microsoft.CSharp" Version="4.5.0" />
+    <PackageReference Update="System.Collections.Immutable" Version="1.3.1" />
+    <PackageReference Include="System.Linq.Async" Version="4.0.0" />
```
We should not have any Include's in this file, only updates. Having this be include will add it to all projects which we don't want.

m-nash replied:
doh, good catch meant to be an update

m-nash committed:
change system.linq.async to update

m-nash committed:
update to use source vs binary for the temporary generation of Azure.ResourceManager.Authorization

m-nash committed:
Add changelog for azure.resourcemanager.authorization and update version to be beta

markcowl commented on sdk/resourcemanager/Proto.Client/Proto-Client.sln:
```diff
@@ -0,0 +1,55 @@
+﻿
```
For parallelism, shouldn't this be Proto.Client.sln?

markcowl commented on sdk/resourcemanager/Proto.Client/authorization/README.MD:
```diff
@@ -0,0 +1,3 @@
+# Proto Authorization
+
+Proto type version of Azure.ResourceManager.Authorization used to showcase and test the new prototype track 2 management plane SDK
```
```suggestion
Prototype version of Azure.ResourceManager.Authorization used to showcase and test the new prototype track 2 management plane SDK
```

markcowl commented on sdk/resourcemanager/Proto.Client/compute/README.MD:
```diff
@@ -0,0 +1,3 @@
+# Proto Compute
+
+Proto type version of Azure.ResourceManager.Compute used to showcase and test the new prototype track 2 management plane SDK
```
```suggestion
Prototype version of Azure.ResourceManager.Compute used to showcase and test the new prototype track 2 management plane SDK
```

markcowl commented on sdk/resourcemanager/Proto.Client/network/README.MD:
```diff
@@ -0,0 +1,3 @@
+# Proto Network
+
+Proto type version of Azure.ResourceManager.Network used to showcase and test the new prototype track 2 management plane SDK
```
```suggestion
Prototype version of Azure.ResourceManager.Network used to showcase and test the new prototype track 2 management plane SDK
```

markcowl commented on sdk/resourcemanager/Azure.ResourceManager.Core/src/Azure.ResourceManager.Core.csproj:
```diff
@@ -0,0 +1,22 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="System.Collections.Immutable" />
```
Does our Azure.Core dependency come through Azure.Identity?

I assume we are not fully integrating into their targets and proeprties files yet.

allenjzhang replied:
We are taking dependency thru build.props that ended up here. [https://github.com/Azure/azure-sdk-for-net/blob/d1e5620b9bce5422892b129de43182507dfc0bdf/sdk/core/Azure.Core/src/Azure.Core.props#L4](https://github.com/Azure/azure-sdk-for-net/blob/d1e5620b9bce5422892b129de43182507dfc0bdf/sdk/core/Azure.Core/src/Azure.Core.props#L4)

---
In reply to: [583344699](https://github.com/Azure/azure-sdk-for-net/pull/18968#discussion_r583344699) [](ancestors = 583344699)

markcowl reviewed (approved):
A couple of minor edits and questions, none of them blocking.

I wonder if we want to have 'core' in the name of the directory inside 'sdk', resourcemanager.core or resourcemanagercore or the like.  I think we can change this easily, so I would not block on the final directory name.

m-nash committed:
update typos in readme

m-nash committed:
update name of proto client sln

allenjzhang commented on sdk/resourcemanager/Azure.ResourceManager.Core/README.md:
```diff
@@ -0,0 +1,77 @@
+# Azure ResourceManager Core client library for .NET
```
Can you open an ADO to come back revisit this readme. The content was originally aimed at RP packages not core.

allenjzhang commented on sdk/resourcemanager/Proto.Client/README.MD:
```diff
@@ -0,0 +1,3 @@
+# Proto Client
```
Wes, what's today's usage of $Root\samples folder? This is like sample client we may want to keep it outside of SDK folder.

weshaggard replied:
The root samples folder is for cross-library samples. If we end up keeping this longer term we should probably move it into a samples folder under Azure.ResourceManager.Core/samples.

m-nash replied:
this will be deleted before moving into master

allenjzhang reviewed (approved):
:shipit:

weshaggard commented on sdk/resourcemanager/Azure.ResourceManager.Core/src/Azure.ResourceManager.Core.csproj:
```diff
@@ -0,0 +1,22 @@
+
+  <ItemGroup>
+    <PackageReference Include="System.Collections.Immutable" />
+    <PackageReference Include="Azure.Identity" />
```
Is there a reason you need to reference Azure.Identity directly? It is my understanding from @schaabs that we need to avoid having direct dependencies on this library and app developers need to be the one that reference it.

m-nash replied:
I added an ADO item to follow up on this, I am not quite sure how to remove this dependency here.

weshaggard reviewed (approved):
I left a few comments, particularly around dependency clean-up that we should follow up on but it doesn't need to block the initial check-in.

m-nash committed:
merge in changes from proto repo this week

m-nash committed:
fixed unit test failures
missed a few files from the merge from proto repo
updated compute and network sdk version to use latest azure.core with readonly list types

m-nash committed:
ignore tests which require live azure until 5122 is closed

m-nash committed:
Update failing scenario test
